### PR TITLE
GODRIVER-1411 fix data race between connection.connect and pool.disconnect

### PR DESCRIPTION
**Pull request by iwysiu**

### CODE REVIEW

File: x/mongo/driver/topology/connection.go [hunk 1]
```diff
@@ -171,6 +176,12 @@ func (c *connection) wait() error {
 	return c.connectErr
 }
 
+func (c *connection) waitForContext() {
+	if c.contextMade != nil {
```
Comment (edaniels) [inline x/mongo/driver/topology/connection.go]: Won't it be the case that this is always non-nil?

File: x/mongo/driver/topology/pool.go [hunk 2]
```diff
@@ -414,6 +412,12 @@ func (p *pool) closeConnection(c *connection) error {
 	delete(p.opened, c.poolID)
 	p.Unlock()
 
+	if atomic.LoadInt32(&c.connected) == connected {
+		c.waitForContext()
+		c.cancelConnect()
```
Comment (edaniels) [inline x/mongo/driver/topology/pool.go]: can collapse these two into an actual cancelConnect function that waits and then cancels

File: x/mongo/driver/topology/pool_test.go [hunk 3]
```diff
@@ -175,6 +175,52 @@ func TestPool(t *testing.T) {
...
+				noerr(t, err)
+				ctx, cancel := context.WithTimeout(context.Background(), 3*time.Microsecond)
+				defer cancel()
+				err = p.disconnect(ctx)
+				noerr(t, err)
+				close(disconnectDone)
+			}()
+			<-getDone
+			close(cleanup)
+			time.Sleep(time.Microsecond)
```
Comment (edaniels) [inline x/mongo/driver/topology/pool_test.go]: Do you need this still?

File: x/mongo/driver/topology/pool_test.go [hunk 4]
```diff
@@ -175,6 +175,52 @@ func TestPool(t *testing.T) {
...
+						_, _ = p.get(getCtx)
+						noerr(t, err)
+						time.Sleep(time.Microsecond)
+					}
+				}
+			}()
+			go func() {
+				_, err := p.get(getCtx)
+				noerr(t, err)
+				ctx, cancel := context.WithTimeout(context.Background(), 3*time.Microsecond)
```
Comment (edaniels) [inline x/mongo/driver/topology/pool_test.go]: Did this allow you to trigger the race with -race? Generally, you have to not use time and instead use channels to coordinate "steps". For instance, you can have each goroutine get into the right state by closing a series of channels in order to force the race, if possible. Without the right internal instrumentation though, you may just need to show that with -count=N, the race happens.

Comment (iwysiu) [reply x/mongo/driver/topology/pool_test.go]: Do you have an example of this? This triggers the race once every few runs, but I'd be happy to switch it to something less hack-y and more consistent.

File: x/mongo/driver/topology/connection.go (hunk 1 above)
Comment (iwysiu) [reply x/mongo/driver/topology/connection.go]: yes, fixed

File: x/mongo/driver/topology/pool.go (hunk 2 above)
Comment (iwysiu) [reply x/mongo/driver/topology/pool.go]: done

File: x/mongo/driver/topology/pool_test.go (hunk 3 above)
Comment (iwysiu) [reply x/mongo/driver/topology/pool_test.go]: done

File: x/mongo/driver/topology/pool_test.go (hunk 4 above)
Comment (edaniels) [reply x/mongo/driver/topology/pool_test.go]: yes! https://github.com/edaniels/go-net-utils/blob/master/track/dialer_test.go#L143

File: x/mongo/driver/topology/connection.go [hunk 5]
```diff
@@ -46,6 +46,8 @@ type connection struct {
 	connectDone      chan struct{}
 	connectErr       error
 	config           *connectionConfig
+	cancelConnect    context.CancelFunc
```
Comment (divjotarora) [inline x/mongo/driver/topology/connection.go]: Can these both be renamed to something more descriptive? e.g. `cancelConnectContext` and `connectContextMade`?

**Pull request merged**